cloudinit/distros/parsers/ifconfig.py

```diff
@@ -0,0 +1,245 @@
+# Copyright(C) 2022 FreeBSD Foundation
+#
+# Author: Mina Galić <me+FreeBSD@igalic.co>
+#
+# This file is part of cloud-init. See LICENSE file for license information.
+
+import copy
+import re
+from functools import lru_cache
+from ipaddress import IPv4Address, IPv4Interface, IPv6Interface
+from typing import Dict, Optional, Tuple
+
+from cloudinit import log as logging
+
+LOG = logging.getLogger(__name__)
+
+
+class Ifstate:
+    """
+    This class holds the parsed state of a BSD network interface.
+    It is itself side-effect free.
+    All methods with side-effects should be implemented on one of the
+    BSDNetworking classes.
+    """
+
+    def __init__(self, name):
+        self.name = name
+        self.inet = {}
+        self.inet6 = {}
+        self.up = False
+        self.options = []
+        self.nd6 = []
+        self.flags = []
+        self.mtu: int = 0
+        self.metric: int = 0
+        self.groups = []
+        self.description: Optional[str] = None
+        self.media: Optional[str] = None
+        self.status: Optional[str] = None
+        self.mac: Optional[str] = None
+        self.macs = []
+        self.vlan = {}
+        self.members = []
+
+    @property
+    def is_loopback(self) -> bool:
+        if "loopback" in self.flags or ("lo" in self.groups):
+            return True
+        return False
+
+    @property
+    def is_physical(self) -> bool:
+        # OpenBSD makes this very easy:
+        if "egress" in self.groups:
+            return True
+        if self.groups == [] and self.media and "Ethernet" in self.media:
+            return True
+        return False
+
+    @property
+    def is_bridge(self) -> bool:
+        if "bridge" in self.groups:
+            return True
+        return False
+
+    @property
+    def is_bond(self) -> bool:
+        if "lagg" in self.groups:
+            return True
+        return False
+
+    @property
+    def is_vlan(self) -> bool:
+        if "vlan" in self.groups or self.vlan:
+            return True
+        return False
+
+    @property
+    def is_wlan(self) -> bool:
+        if "wlan" in self.groups or self.vlan:
+            return True
+        return False
+
+
+# see man ifconfig(8)
+# - https://man.freebsd.org/ifconfig(8)
+# - https://man.netbsd.org/ifconfig.8
+# - https://man.openbsd.org/ifconfig.8
+class Ifconfig:
+    def __init__(self):
+        self._ifs = {}
+
+    @lru_cache()
+    def parse(self, text: str) -> Dict[str, Ifstate]:
+        """
+        parse returns a dict of Ifstates
+        This dict will always be the same, given the same input, so we can
+        lru_cache() it. note that lru_cache() takes only the __hash__() of the
+        input (text), so it should be fairly quick, despite our giant inputs.
+        """
+        for line in text.splitlines():
+            if len(line) == 0:
+                continue
+            if line[0] not in ("\t", " "):
```

Comment:
maybe, but: What is the default value for `curif` and `dev`?

`curif` could be `''`, which is a valid key in a `dict`, so I'd have to check that…

but for `dev`, the only sensible default is None, and then `mypy` gets really petty about all the assignments to `dev[x]`, because they are now all `Optional`.

Reply:
I've added a comment to explain what's going on here, and hope it helps
because initializing these at the top of the loop is impossible.